config/html-webpack-inline-critical-css-plugin.js

```diff
@@ -0,0 +1,255 @@
+  nodeName: {
+    get: function() {
+      return this.tagName;
+    }
+  },
+  appendChild: function (child) {
+    treeUtils.appendChild(this, child);
+    return child;
+  },
+  removeChild: function (child) {
+    treeUtils.detachNode(child);
+  },
+  setAttribute: function (name, value) {
+    if (this.attribs == null) this.attribs = {};
+    if (value == null) value = '';
+    this.attribs[name] = value;
+  },
+  removeAttribute: function(name) {
+    if (this.attribs != null) {
+      delete this.attribs[name];
+    }
+  },
+  getAttribute: function (name) {
+    return this.attribs != null && this.attribs[name];
+  },
+  hasAttribute: function (name) {
+    return this.attribs != null && this.attribs[name] != null;
+  },
+  getAttributeNode: function (name) {
+    const value = this.getAttribute(name);
+    if (value!=null) return { specified: true, value: value };
+  },
+  getElementsByTagName: getElementsByTagName
+};
+
+const DocumentExtensions = {
+  nodeType: {
+    get: function () {
+      return 11;
+    }
+  },
+  createElement: function (name) {
+    return treeUtils.createElement(name, null, []);
+  },
+  createTextNode: function (text) {
+    const scratch = this.$$scratchElement;
+    treeUtils.insertText(scratch, text);
+    const node = scratch.lastChild;
+    treeUtils.detachNode(node);
+    return node;
+  },
+  querySelector: function (sel) {
+    return this.$match.first(sel);
+  },
+  querySelectorAll: function (sel) {
+    return this.$match.select(sel);
+  },
+  getElementsByTagName: getElementsByTagName,
+  // https://github.com/dperini/nwmatcher/blob/3edb471e12ce7f7d46dc1606c7f659ff45675a29/src/nwmatcher.js#L353
+  addEventListener: Object
+};
+
+module.exports = class HtmlWebpackInlineCriticalCssPlugin {
+  constructor(options) {
+    this.options = options || {};
+  }
+
+  apply(compiler) {
+    const self = this;
+    const outputPath = compiler.options.output.path;
+    compiler.hooks.compilation.tap(PLUGIN_NAME, function (compilation) {
+      compilation.hooks.htmlWebpackPluginAfterHtmlProcessing.tapAsync(PLUGIN_NAME, function (htmlPluginData, callback) {
+        const document = parse5.parse(htmlPluginData.html, PARSE5_OPTS);
+
+        defineProperties(document, DocumentExtensions);
+        document.documentElement = document.childNodes[document.childNodes.length - 1];
+
+        const scratch = document.$$scratchElement = document.createElement('div');
+        const elementProto = Object.getPrototypeOf(scratch);
+        defineProperties(elementProto, ElementExtensions);
+        elementProto.ownerDocument = document;
+
+        document.$match = nwmatcher({ document });
+        document.$match.configure({
+          CACHING: false,
+          USE_QSAPI: false,
+          USE_HTML5: false
+        });
+
+        const externalSheets = document.querySelectorAll('link[rel="stylesheet"]');
+
+        Promise.all(externalSheets.map(function(link) {
+          const href = link.getAttribute('href');
```

Comment:
Use `link.href` to always get a normalized URL string (even when the attribute value is a relative URL, for example).

Reply:
nah, this DOM (parse5) is extremely rudimentary and only supports attributes.

Reply:
Ooh, gotcha. Fingers crossed dealing with raw attribute values doesn’t break anything, I guess!